File: solutions/127.word-ladder.java
/*
 * @lc app=leetcode id=127 lang=java
 *
 * [127] Word Ladder
 */

// @lc code=start
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> dict = new HashSet<>(wordList);
        
        Queue<String> que = new LinkedList<>();
        que.offer(beginWord);
        
        int step = 0;
        
        while(!que.isEmpty()) {
            step += 1;
            int n = que.size();
            
            for(int i=0; i<n; ++i) {
                var word = que.poll();
                if(word.equals(endWord)) return step; //trap 不要用 word==endWord
                
                var charArr = word.toCharArray();
                for(int j=0; j<charArr.length; ++j) {
                    char oriChar = charArr[j];
                    for(var c='a'; c<='z'; ++c) {
                        charArr[j] = c;
                        var newWord = new String(charArr);
                        if(dict.contains(newWord)) {
                            que.offer(newWord);
                            dict.remove(newWord);
                        }
                    }
                    charArr[j] = oriChar;
                }
            }
        }
        
        return 0;
    }
}
// @lc code=end

File: solutions/127.word-ladder.rb
#
# @lc app=leetcode id=127 lang=ruby
#
# [127] Word Ladder
#

# @lc code=start
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {Integer}
def ladder_length(begin_word, end_word, word_list)
  word_set = Set.new(word_list)
  return 0 unless word_set.include? end_word

  q = [begin_word]
  step = 0
  until q.empty?
    step += 1
    q.size.times do
      word = q.shift
      return step if word == end_word

      (0...word.length).each do |i|
        ('a'..'z').each do |c|
          new_word = word.dup
          new_word[i] = c
          if word_set.include? new_word
            q << new_word
            word_set.delete new_word
          end
        end
      end
    end
  end
  0
end
# @lc code=end

ladder_length 'hit',
              'cog', %w[hot dot dog lot log cog]
